runtime/gc_base/ContinuationObjectList.hpp

```diff
@@ -0,0 +1,152 @@
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#ifndef CONTINUATIONOBJECTLIST_HPP_
+#define CONTINUATIONOBJECTLIST_HPP_
+
+#include "j9.h"
+#include "j9cfg.h"
+#include "modron.h"
+
+#include "BaseNonVirtual.hpp"
+#include "ObjectAccessBarrier.hpp"
+
+class MM_EnvironmentBase;
+
+/**
+ * A global list of Continuation objects.
+ */
+class MM_ContinuationObjectList : public MM_BaseNonVirtual
+{
+/* data members */
+private:
+	volatile j9object_t _head; /**< the head of the linked list of Continuation objects */
+	j9object_t _priorHead; /**< the head of the linked list before Continuation object processing */
+	MM_ContinuationObjectList *_nextList; /**< a pointer to the next Continuation list in the global linked list of lists */
+	MM_ContinuationObjectList *_previousList; /**< a pointer to the previous Continuation list in the global linked list of lists */
+#if defined(J9VM_GC_VLHGC)
+	uintptr_t _objectCount; /**< the number of objects in the list */
+#endif /* defined(J9VM_GC_VLHGC) */
+protected:
+public:
+
+/* function members */
+private:
+protected:
+public:
+	/**
+	 * Add the specified linked list of objects to the buffer.
+	 * The objects are expected to be in a NULL terminated linked
+	 * list, starting from head and end at tail.
+	 * This call is thread safe.
+	 *
+	 * @param env[in] the current thread
+	 * @param head[in] the first object in the list to add
+	 * @param tail[in] the last object in the list to add
+	 */
+	void addAll(MM_EnvironmentBase* env, j9object_t head, j9object_t tail);
+
+	/**
+	 * Fetch the head of the linked list, as it appeared at the beginning of Continuation object processing.
+	 * @return the head object, or NULL if the list is empty
+	 */
+	MMINLINE j9object_t getHeadOfList() { return _head; }
+
+	/**
+	 * Move the list to the prior list and reset the current list to empty.
+	 * The prior list may be examined with wasEmpty() and getPriorList().
+	 */
+	void startContinuationProcessing() {
+		_priorHead = _head;
+		_head = NULL;
+#if defined(J9VM_GC_VLHGC)
+		clearObjectCount();
+#endif /* defined(J9VM_GC_VLHGC) */
+	}
+
+	/**
+	 * copy the list to the prior list for backup, can be restored via calling backoutList()
+	 */
+	MMINLINE void backupList() {
+		_priorHead = _head;
+	}
+
+	/**
+	 * restore the list from the prior list, and reset the prior list to empty.
+	 */
+	MMINLINE void backoutList() {
+		_head = _priorHead;
```

Comment:
Hm, don't we need to do some real work here (walking the list and reverting fixed-up slot), like Unfinalized and similar does? But, we can deal with it in a follow-up change (leave a TODO here).

Reply:
processing Continuation list will remove  dead one from the list and clean up related native structure J9VMContinuation, after clean-up, I am not sure that we can rollback, but anyway I don't think we need  J9VMContinuation for the dead object after backout, so as long as keeping vmref (point to J9VMContinuation) in Continuation Object as NULL, we would not have any issue.